### PR TITLE
1284 show theorem and proof attempt state

### LINKED ISSUE #1284: Show theorem/proof attempt state
As soon as #1256 is completed, add the state of theorems and proof attempts to the views. Show it dynamically like the ontology version state.

### PR DESCRIPTION
**Pull request by eugenk**
This shall fix #1284. It also adds retry functionality to failed proof attempts.

~~The branch of this pull request is based on:~~
* ~~add_tactic_script_model (#1311)~~
* ~~1256-add_proving_related_serializers (#1289)~~

~~The first commit of this branch is:~~
~~Move state helper to own module and generalize it.	33f017c~~

### CODE REVIEW

**Comment by eugenk** [comment]
I rebased this on staging.

**Comment by tillmo** [comment]
when trying this with legacy data, I get an error in `app/views/proof_attempts/index.html.haml, line 26`:
```
NoMethodError at /test/Family/FamilyBase/Ax1_19/proof_attempts
undefined method `gsub' for nil:NilClass
```
Details at https://gist.github.com/tillmo/7a3bcb2b74c96fbc718b

**Comment by eugenk** [comment]
That's odd. I can't find any occurrence of `gsub` in that line (app/controllers/application_controller.rb, line 67).

**Comment by eugenk** [comment]
I can't reproduce this error using data that I created on staging.ontohub.org. This branch is a little older than staging. Could you open the rails console and execute the following, please:
```ruby
Theorem.find_with_locid('/test/Family//FamilyBase//Ax1_19').proof_attempts.each { |pa| puts pa.inspect; puts }
```
What is being printed to the console?

**Comment by tillmo** [comment]
sorry, that line was the botton of the stack trace. The top is ` app/controllers/application_controller.rb, line 67 `

**Comment by tillmo** [comment]
Here is the console output:
```
[1] pry(main)> Theorem.find_with_locid('/test/Family//FamilyBase//Ax1_19').proof_attempts.each { |pa| puts pa.inspect; puts }
  Theorem Load (0.7ms)  SELECT "sentences".* FROM "sentences" WHERE "sentences"."imported" = 'f' AND "sentences"."type" IN ('Theorem') AND "sentences"."locid" = '/test/Family//FamilyBase//Ax1_19' LIMIT 1
  ProofAttempt Load (0.4ms)  SELECT "proof_attempts".* FROM "proof_attempts" WHERE "proof_attempts"."sentence_id" = 1405
#<ProofAttempt id: 39, proof_status_id: "UNK", time_taken: 0, sentence_id: 1405, created_at: "2015-03-22 09:32:06", updated_at: "2015-03-22 09:32:06", number: 1, prover_id: 6, state: "pending", state_updated_at: nil, last_error: nil, proof_attempt_configuration_id: nil, locid: nil>

#<ProofAttempt id: 65, proof_status_id: "THM", time_taken: 0, sentence_id: 1405, created_at: "2015-04-28 06:01:05", updated_at: "2015-04-28 06:01:07", number: 2, prover_id: 13, state: "done", state_updated_at: "2015-04-28 06:01:07", last_error: nil, proof_attempt_configuration_id: 16, locid: "/test/Family//FamilyBase//Ax1_19//proof-attempt-2">

=> [#<ProofAttempt id: 39, proof_status_id: "UNK", time_taken: 0, sentence_id: 1405, created_at: "2015-03-22 09:32:06", updated_at: "2015-03-22 09:32:06", number: 1, prover_id: 6, state: "pending", state_updated_at: nil, last_error: nil, proof_attempt_configuration_id: nil, locid: nil>,
 #<ProofAttempt id: 65, proof_status_id: "THM", time_taken: 0, sentence_id: 1405, created_at: "2015-04-28 06:01:05", updated_at: "2015-04-28 06:01:07", number: 2, prover_id: 13, state: "done", state_updated_at: "2015-04-28 06:01:07", last_error: nil, proof_attempt_configuration_id: 16, locid: "/test/Family//FamilyBase//Ax1_19//proof-attempt-2">]
[2] pry(main)> 
```

**Comment by eugenk** [comment]
Please run `rake data:migrate` and check if the problem still exists.

**Comment by tillmo** [comment]
I get
```
[2] pry(main)> quit
till@mark:~/web2.0/ontohub$ rake data:migrate
==  AddStateToProofAttemptsAndTheorems: migrating =============================
==  AddStateToProofAttemptsAndTheorems: migrated (0.0714s) ====================

==  CreateProofAttemptConfigurations: migrating ===============================
==  CreateProofAttemptConfigurations: migrated (0.4015s) ======================

==  AddLocIdToProofAttempt: migrating =========================================
==  AddLocIdToProofAttempt: migrated (0.1622s) ================================

==  AddLocIdToProofAttemptConfiguration: migrating ============================
rake aborted!
StandardError: An error has occurred, this and all later migrations canceled:

undefined method `locid' for nil:NilClass/home/till/web2.0/ontohub/app/models/proof_attempt_configuration.rb:32:in `generate_locid'
/home/till/web2.0/ontohub/db/data/20150325061425_add_loc_id_to_proof_attempt_configuration.rb:5:in `block in up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:21:in `block (2 levels) in find_each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:21:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:21:in `block in find_each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:72:in `find_in_batches'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:20:in `find_each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/querying.rb:8:in `find_each'
/home/till/web2.0/ontohub/db/data/20150325061425_add_loc_id_to_proof_attempt_configuration.rb:3:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:410:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:410:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:129:in `with_connection'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:389:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:528:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:720:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `call'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `block in ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/transactions.rb:208:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:719:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:570:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:551:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/data_migrate-1.2.0/tasks/databases.rake:221:in `block (2 levels) in <top (required)>'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `eval'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `<main>'
NoMethodError: undefined method `locid' for nil:NilClass
/home/till/web2.0/ontohub/app/models/proof_attempt_configuration.rb:32:in `generate_locid'
/home/till/web2.0/ontohub/db/data/20150325061425_add_loc_id_to_proof_attempt_configuration.rb:5:in `block in up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:21:in `block (2 levels) in find_each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:21:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:21:in `block in find_each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:72:in `find_in_batches'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/relation/batches.rb:20:in `find_each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/querying.rb:8:in `find_each'
/home/till/web2.0/ontohub/db/data/20150325061425_add_loc_id_to_proof_attempt_configuration.rb:3:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:410:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:410:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/connection_pool.rb:129:in `with_connection'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:389:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:528:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:720:in `block (2 levels) in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `call'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `block in ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/connection_adapters/abstract/database_statements.rb:192:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/transactions.rb:208:in `transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:775:in `ddl_transaction'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:719:in `block in migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `each'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:700:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:570:in `up'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/activerecord-3.2.21/lib/active_record/migration.rb:551:in `migrate'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/gems/data_migrate-1.2.0/tasks/databases.rake:221:in `block (2 levels) in <top (required)>'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `eval'
/home/till/.rvm/gems/ruby-2.1.5@ontohub/bin/ruby_executable_hooks:15:in `<main>'
Tasks: TOP => data:migrate
(See full trace by running task with --trace)
```

**Comment by eugenk** [comment]
Alright. It seems like I need to redo the data migrations rather sooner than later.
I will create another pull request today that fixes the data migrations and needs to be merged before the review here can continue.

**Comment by tillmo** [comment]
maybe the problem has been caused by the fact that I was not disciplined enough to run the data migrations each time, and therefore db and data migrations have been exectued in wrong order?

**Comment by eugenk** [comment]
Even if that caused the problem, this should not happen. I'm fixing the data migrations right now.

**Comment by eugenk** [comment]
#1335 should fix the migration problem.

**Comment by tillmo** [comment]
OK, my problem has been solved by #1335 (although it still occurs here, but I assume that I can safely ignore that).

File: app/controllers/ontologies_controller.rb [hunk 1]
```diff
@@ -13,9 +13,6 @@ class OntologiesController < InheritedResources::Base
 
   respond_to :html, except: %i(show)
 
-  before_filter :check_write_permission, except: [:index, :show, :oops_state]
-  before_filter :check_read_permissions
```
Comment (tillmo) [inline app/controllers/ontologies_controller.rb]: why are these `before_filters` removed?

**Comment by eugenk** [comment]
I just rebased this on staging and amended the odd commit removing the before filters.

File: app/controllers/ontologies_controller.rb (hunk 1 above)
Comment (eugenk) [inline app/controllers/ontologies_controller.rb]: This is odd. It should not have happened. I probably messed something up with my editor without noticing it. This change is removed (by commit amend) in 318de55.

**Comment by eugenk** [comment]
Rebased again. I made a mistake in the last rebase. The amended commit is now 06cbff1e355ede78f5fa530b57a93e4dbc533840.

**Comment by tillmo** [comment]
:+1:

**Pull request merged**